### PR TITLE
Added TResult generic parameter to BindOnFailure and BindOnFailureAsync

### LINKED ISSUE #102: Failure type mapping for BindOnFailure()
For the cases when I want to "first try Method A, and if that fails try Method B":

``` c#
Result<Success, FailureA> MethodA() { ... }
Result<Success, FailureB> MethodB() { ... }

Result<Success, FailureB> result = MethodA().BindOnFailure(error => MethodB());
```

This seems to be problematic given the current signature of `BindOnFailure()`:
``` c#
public static Result<TSuccess, TFailure> BindOnFailure<TSuccess, TFailure>(
    this Result<TSuccess, TFailure> result, 
    Func<TFailure, Result<TSuccess, TFailure>> bind
)
```

However the current `Bind()` signature allows for type mapping the success side:
``` c#
public static Result<TResult, TFailure> Bind<TSuccess, TFailure, TResult>(
    this Result<TSuccess, TFailure> result, 
    Func<TSuccess, Result<TResult, TFailure>> bind
)
```

It seems like a straightforward solution would be to change the `BindOnFailure()` signature to:
``` c#
public static Result<TSuccess, TFailure> BindOnFailure<TSuccess, TFailure, TResult>(
    this Result<TSuccess, TFailure> result, 
    Func<TFailure, Result<TSuccess, TResult>> bind
)
```

This would technically be a breaking change (although most syntactic changes should be automagically handled through generic type inferencing).

Note: my current work-around is to `Transpose()` a few times around a `Bind()`, which to me speaks to the missed intention of `BindOnFailure()`:

``` c#
Result<Success, FailureB> result = MethodA().Transpose().Bind(error => MethodB().Transpose()).Transpose();
```

### PR DESCRIPTION
**Pull request by deimors**
Resolves #102 

The change wasn't breaking as far as the tests are concerned, I only updated them to demonstrate the `TFailure` -> `TResult` type mapping, however this is a breaking change in the strictest sense (ie: if generic parameters are explicit). @JohannesMoersch thoughts on what the new version number should be?

Because the change is a strict generalization of the existing method, I chose to modify the existing tests rather than introduce new ones.

I also left `BindIfFailure()` alone, as it's already marked as obsolete.

### CODE REVIEW

File: Functional.Primitives.Extensions/ResultAsyncExtensions.cs [hunk 1]
```diff
@@ -79,16 +79,16 @@ public static partial class ResultAsyncExtensions
 		public static async Task<Result<TResult, TFailure>> BindAsync<TSuccess, TFailure, TResult>(this Task<Result<TSuccess, TFailure>> result, Func<TSuccess, Task<Result<TResult, TFailure>>> bind)
 			=> await (await result).BindAsync(bind);
 
-		public static async Task<Result<TSuccess, TFailure>> BindOnFailureAsync<TSuccess, TFailure>(this Result<TSuccess, TFailure> result, Func<TFailure, Task<Result<TSuccess, TFailure>>> bind)
+		public static async Task<Result<TSuccess, TResult>> BindOnFailureAsync<TSuccess, TFailure, TResult>(this Result<TSuccess, TFailure> result, Func<TFailure, Task<Result<TSuccess, TResult>>> bind)
```
Comment (NathanMagnus) [inline Functional.Primitives.Extensions/ResultAsyncExtensions.cs:82]: Is there a reason this was changed from `TFailure` to `TResult`?

Comment (NathanMagnus) [reply Functional.Primitives.Extensions/ResultAsyncExtensions.cs:82]: Oh scratch that, I totally misread the above line.

**Comment by RyanMarcotte** [comment]
Requesting to merge and release #101 before this if a major version number upgrade will be done as part of this PR.

**Comment by deimors** [comment]
Yea @RyanMarcotte that makes sense to me 👍

**Comment by JohannesMoersch** [comment]
While for anyone compiling fresh, the type inferencing will probably cover them, any already compiled code (like in a library) using that extension will blow up. Can you leave the old versions please (and mark them with [Obsolete] and [EditorBrowsable(EditorBrowsableState.Never)])? We'll just need to make sure the compiler can still choose the correct extension when both exist.

**Comment by JohannesMoersch** [comment]
If everything looks good after that, can you bump third octet (since it's really a bug fix?) please?

**Comment by deimors** [comment]
What's your thoughts on how to make sure the compiler chooses the right extension? Right away I'm running into that ambiguity problem in the tests. Apart from different naming or always supplying the generic parameters, I'm not what other options there are.

**Comment by JohannesMoersch** [comment]
You know what might work? Make the old method not an extension method. Just remove the "this". It would still have the same signature as before, so anything compiled against it should still work. We might need to test this concept in a test project though. I could try that later.

**Comment by deimors** [comment]
Ah I think I get ya, the extension method syntax is just there to build the AST when compiling but shouldn't matter at all to the IL, it's just a method call at that point. I'll give that a shot and get something up soon.

**Review by JohannesMoersch — DISMISSED** [review]

**Comment by JohannesMoersch** [comment]
Oh, crap. I've installed the GitHub app now so I'm hoping it'll ping me when someone opens a PR here or comments.

**Review by JohannesMoersch — APPROVED** [review]

**Pull request merged**